library/src/scala/deriving.scala

```diff
@@ -0,0 +1,69 @@
+package scala
+
+object deriving {
+
+  /** Mirrors allows typelevel access to enums, case classes and objects, and their sealed parents.
+   */
+  sealed trait Mirror {
+
+    /** The mirrored *-type */
+    type MonoType
+  }
+
+  object Mirror {
+
+    /** The Mirror for a sum type */
+    trait Sum extends Mirror { self =>
+
+      type ElemTypes <: Tuple
+
+      /** The ordinal number of the case class of `x`. For enums, `ordinal(x) == x.ordinal` */
+      def ordinal(x: MonoType): Int
+    }
+
+    /** The Mirror for a product type */
+    trait Product extends Mirror {
+
+      /** The types of the elements */
+      type ElemTypes <: Tuple
+
+      /** The name of the whole product type */
+      type CaseLabel <: String
+
+      /** The names of the product elements */
+      type ElemLabels <: Tuple
+
+      /** Create a new instance of type `T` with elements taken from product `p`. */
+      def fromProduct(p: scala.Product): MonoType
+    }
+
+    trait Singleton extends Product {
```

Comment:
How is this different from an empty product? Why do we need such distinction?

Reply:
First, it's an optimization since we do not need to spell out ElemTypes and CaseLabels.
Second, it's important for printing. A case class `case class C()` should print as `C()`, but
a case object `case object C` should print as `C`. We can tell the difference since the mirror
of the case object is a `Mirror.Singleton`.